dex/farm/src/custom_rewards.rs

```diff
@@ -1,54 +1,121 @@
 elrond_wasm::imports!();
 elrond_wasm::derive_imports!();
 
+const BLOCKS_IN_YEAR: u64 = 31_536_000 / 6; // seconds_in_year / 6_seconds_per_block
+
 #[elrond_wasm::module]
 pub trait CustomRewardsModule:
     config::ConfigModule
     + token_send::TokenSendModule
     + farm_token::FarmTokenModule
     + rewards::RewardsModule
 {
-    fn mint_per_block_rewards(&self, token_id: &TokenIdentifier) -> BigUint {
+    fn calculate_extra_rewards_since_last_allocation(&self) -> BigUint {
         let current_block_nonce = self.blockchain().get_block_nonce();
         let last_reward_nonce = self.last_reward_block_nonce().get();
 
         if current_block_nonce > last_reward_nonce {
-            let to_mint = self.calculate_per_block_rewards(current_block_nonce, last_reward_nonce);
-
-            if to_mint != 0 {
-                self.send().esdt_local_mint(token_id, 0, &to_mint);
-            }
+            let extra_rewards =
+                self.calculate_per_block_rewards(current_block_nonce, last_reward_nonce);
             self.last_reward_block_nonce().set(&current_block_nonce);
-            to_mint
+
+            extra_rewards
         } else {
             BigUint::zero()
         }
     }
 
-    fn generate_aggregated_rewards(&self, reward_token_id: &TokenIdentifier) {
-        let total_reward = self.mint_per_block_rewards(reward_token_id);
-        if total_reward > 0 {
-            self.increase_reward_reserve(&total_reward);
-            self.update_reward_per_share(&total_reward);
+    fn generate_aggregated_rewards(&self) {
+        let extra_rewards = self.calculate_extra_rewards_since_last_allocation();
+        if extra_rewards > 0 {
+            self.increase_reward_reserve(&extra_rewards);
+            self.update_reward_per_share(&extra_rewards);
         }
     }
 
+    fn calculate_rewards_with_apr_limit(
+        &self,
+        amount: &BigUint,
+        current_reward_per_share: &BigUint,
+        initial_reward_per_share: &BigUint,
+    ) -> BigUint {
+        let unbounded_rewards =
+            self.calculate_reward(amount, current_reward_per_share, initial_reward_per_share);
+        if unbounded_rewards == 0u32 {
+            return unbounded_rewards;
+        }
+
+        let farming_token_total_liquidity = self.farming_token_total_liquidity().get();
+        let max_apr = self.max_annual_percentage_rewards().get();
+        let max_rewards = farming_token_total_liquidity * max_apr
```

Comment:
not sure about this calculus :thinking: maybe we'll reconsider this together

Reply:
Fixed in latest commit.